middleware/proxy.go

```diff
@@ -29,6 +29,19 @@ type (
 		// Required.
 		Balancer ProxyBalancer
 
+		// RetryCount defines the number of times a proxied request to an unavailable
+		// ProxyTarget should be retried using the next available ProxyTarget. Defaults
+		// to 0, meaning requests are never retried.
+		RetryCount int
+
+		// RetryFilter defines a function used to determine if a failed request to an
+		// unavailable ProxyTarget should be retried. The RetryFilter will only be called
+		// when the number of previous retries is less than RetryCount. If the function returns
+		// true, the request will be retried. When not specified, DefaultProxyRetryFilter
+		// will be used, which will always retry requests. A user defined ProxyRetryFilter
+		// can be provided to only retry specific requests, for example only retry GET requests.
+		RetryFilter ProxyRetryFilter
```

Comment:
I propose `ErrorHandler func(err error, c echo.Context) error` instead of this.  This would open new use-cases for users - for example - customize error codes and messages that are returned.

also these function types `ProxyRetryFilter func(c echo.Context) bool` are in my personal opinion annoying as when I check what that field is - I do not see immediately that it is function with following arguments. Basically it is harder to understand code and only IDEs features like onhover tooltips help you here. For these kind, one-off fields, this is not good user experience IMO.  

Defining new types for middleware config handler/func fields does not give anything except new public type that we must support due semantic versioning.

Reply:
Agreed on the function types - I was mostly trying to stick to the conventions of the existing code (funtion types seem to be used quite a lot) but I agree with your points, have changed this.

> - for example - customize error codes and messages that are returned.

I think that is a valid use case, but I also think that should be a seperate function from the proxy filter. I feel that having 1 function to serve 2 purposes ("should this request be retried?" and "what error should be returned to clients?") could be confusing for users. Imagine the scenario where a user wants to retry failed requests, but then when the request can no longer be retried (we run out of retry attepts) they want to customise the error code or message. With this proposal, I don't think it is possible to implement this. It might also be confusing that the function can return nil, but then we don't retry anyway as we've run out of retry attempts. It also feels a bit of a messy interface if you have to return `nil` to indicate you want something retried (but that we might not retry anyway).

How would you feel about adding `ErrorHandler` as seperate function on the ProxyConfig? It's purpose is to allow users to customise error responses and messages, unrelated to retry? The RetryFilter would work as it does now (I'll reply re: the error on your other comment). That way the seperation is clear. To customise what requests are retried, provide a RetryFilter. To customise errors that are returned to the clientm provide a ErrorHandler.

Reply:
alright, lets keep `RetryFilter`.  I think `RetryFilter` benefit if it would take err in. Reasoning is that retry is most likely based on what that error is and getting it from context is annoying - you need cast etc.

also lets add `ErrorHandler` as additional function.